src/modalities/models/components/collator_wrapper.py

```diff
@@ -0,0 +1,113 @@
+        collate_fn: CollateFnIF,
+        target_keys_to_mask: List[str],
+        loss_ignore_index: int,
+        mask_tokens: MaskingTokenConfig,
+        tokenizer: TokenizerWrapper,
+    ):
+        """Wraps the given collate_fn and masks the target keys if not within the given special mask tokens.
+        Does not include both mask tokens into the loss. If you need a token to indicate the end of the assistant,
+        use another special token for this!
+        Works also for the continuous dataset reading, as if the "end-include-to-loss" token is detected in the front,
+        all tokens before are included to the loss.
+
+        Throws a ValueError if the mask tokens are not found in the target or if the mask tokens are the same.
+        """
+        self.collate_fn = collate_fn
+        self.target_keys_to_mask = target_keys_to_mask
+        self.loss_ignore_index = loss_ignore_index
+        self.tokenizer = tokenizer
+        self.b_mask_token_id = self.tokenizer.get_token_id(mask_tokens.b_include_to_loss_token)
+        self.e_mask_token_id = self.tokenizer.get_token_id(mask_tokens.e_include_to_loss_token)
+        if self.b_mask_token_id == self.e_mask_token_id:
+            raise ValueError(
+                "b_mask_token_id and e_mask_token_id of the " + "LossMaskingCollateFnWrapper must be different!"
+            )
+
+    def __call__(self, batch: List[Dict[str, torch.Tensor]]) -> DatasetBatch:
+        dataset_batch = self.collate_fn(batch)
+        for target_key_to_mask in self.target_keys_to_mask:
+            target = dataset_batch.targets[target_key_to_mask]
+            masked_target = self._mask_target(
+                target=target,
+                b_mask_token_id=self.b_mask_token_id,
+                e_mask_token_id=self.e_mask_token_id,
+                loss_ignore_index=self.loss_ignore_index,
+            )
+            dataset_batch.targets[target_key_to_mask] = masked_target
+        return dataset_batch
+
+    def _mask_target(
+        self, target: torch.Tensor, b_mask_token_id: int, e_mask_token_id: int, loss_ignore_index: int
+    ) -> torch.Tensor:
+        error_msg = ""
+        if b_mask_token_id not in target:
+            error_msg += "b_mask_token_id not found in target."
+        if e_mask_token_id not in target:
+            error_msg += "e_mask_token_id not found in target."
+        if error_msg:
+            raise ValueError(
+                f"{error_msg} in masking tokens for loss computation. "
+                + "Make sure the tokenizer tokenizes as expected. "
+                + "Frequent source of error is the tokenization of spaces: "
+                + "e.g. ' <token>' and '<token>' are different tokens."
+            )
+
+        mask = torch.zeros_like(target)
+        # we shift the mask to the right, to exclude not only the end mask token but also
+        # the begin mask token from the loss
+        mask[:, 1:] += torch.where(target != b_mask_token_id, 0, 1)[:, :-1]
```

Comment:
I added the following comment within the code:
```python
"""
We mask the target tensor with loss_ignore_index between, but not inclusive the begin and end mask token.
We do this vectorizes, as this is fast.
Example:
    sample_orig =      [2,2,3,2, 2,4,2,2,2]
    sample =           [2,2,3,2, 2,4,2,2] # from collate_fn
    target =           [2,3,2,2, 4,2,2,2] # from collate_fn
    mask_initially =   [0,0,0,0, 0,0,0,0] # mask = torch.zeros_like(target)
    mask_shifted_1 =   [0,0,1,0, 0,0,0,0] # mask[:, 1:] += torch.where(target != b_mask_token_id, 0, 1)[:, :-1]
    mask_shifted_2 =   [0,0,1,0,-1,0,0,0] # mask += torch.where(target != e_mask_token_id, 0, -1)
    mask_cumsum =      [0,0,1,1, 0,0,0,0] # include_to_loss_mask = mask.cumsum(-1)

By shifting only the b_mask_token_id to the right, we exclude the end mask token from the loss, as otherwise
cumsum would include the end mask token. E.g. not shifted: [0,1,1,1,0,0,0,0] <- cumsum([0,1,0,0,-1,0,0,0])
"""
```

I changed also  
"exclude not only the end mask token but also the begin mask token from the loss" to 
"exclude not only the begin mask token but also the end mask token from the loss"